# Find subjects for local explanation analyses

Criterium: subjects should have no missing values in the demographic / health information of interest

In [1]:
import pandas as pd
from brain_age_prediction import utils

In [2]:
# save path for predictions
overview_path = '/ritter/share/projects/laura_riedel_thesis/predictions/predictions_overview_100-500p.csv'
# load data overview
data_overview_full = pd.read_csv(overview_path)
# limit to IDs present in heldout test set
data_overview_heldout = data_overview_full[data_overview_full['split']=='heldout_test'].copy()
data_overview_heldout.reset_index(inplace=True,drop=True)
data_overview_heldout = utils.detrend_bag(data_overview_heldout, models=['orig','new'])

# exclude rows with nans
no_nans = ((~data_overview_heldout['bmi'].isnull())
           & (~data_overview_heldout['digit substitution'].isnull())
           & (~data_overview_heldout['education'].isnull())
           & (~data_overview_heldout['fluid intelligence'].isnull())
           & (~data_overview_heldout['grip'].isnull())
           & (~data_overview_heldout['depressive episode'].isnull())
           & (~data_overview_heldout['all depression'].isnull())
           & (~data_overview_heldout['recurrent depressive disorder'].isnull())
           & (~data_overview_heldout['multiple sclerosis'].isnull())
           & (~data_overview_heldout['sex'].isnull())
           & (~data_overview_heldout['weekly beer'].isnull())
           & (~data_overview_heldout['genetic pc 1'].isnull())
           & (~data_overview_heldout['genetic pc 2'].isnull())
           & (~data_overview_heldout['genetic pc 3'].isnull()))
# define columns of interest
columns = ['eid', 'age'] + [col for col in data_overview_heldout.columns[3:-8]] + [col for col in data_overview_heldout.columns[-4:]]

In [3]:
# very young
data_overview_heldout[no_nans].sort_values(by='age', ascending=True)[columns].head(10)

eid  age      bmi  digit substitution  education  \
1491  1137757   48  25.5993                20.0       20.0   
632   1058349   49  26.8186                18.0       20.0   
4234  1389126   49  31.9020                21.0       20.0   
1582  1144954   49  27.3821                24.0       20.0   
272   1025552   49  28.5813                22.0       20.0   
1825  1169644   49  24.2240                29.0       20.0   
3416  1317950   50  21.2939                22.0       10.0   
936   1089320   50  26.8359                18.0       20.0   
1354  1125280   50  32.8291                25.0       15.0   
1730  1159744   50  26.6385                23.0       20.0   

      fluid intelligence  grip  depressive episode  all depression  \
1491                 7.0  50.0                 0.0             0.0   
632                  3.0  26.0                 0.0             0.0   
4234                 6.0  36.0                 0.0             0.0   
1582                 9.0  57.0                 0.0             0.0   
272                  5.0  44.0                 0.0             0.0   
1825                 9.0  26.0                 0.0             0.0   
3416                 5.0  31.0                 0.0             0.0   
936                  5.0  62.0                 0.0             0.0   
1354                 5.0  53.0                 0.0             0.0   
1730                 9.0  58.0                 0.0             0.0   

      recurrent depressive disorder  multiple sclerosis  sex  weekly beer  \
1491                            0.0                 0.0    1          3.0   
632                             0.0                 0.0    0          5.0   
4234                            0.0                 0.0    1         20.0   
1582                            0.0                 0.0    1          2.0   
272                             0.0                 0.0    1          4.0   
1825                            0.0                 0.0    1          4.0   
3416                            0.0                 0.0    0          0.0   
936                             0.0                 0.0    1          5.0   
1354                            0.0                 0.0    1          3.0   
1730                            0.0                 0.0    1         10.0   

      genetic pc 1  genetic pc 2  genetic pc 3   bag_orig    bag_new  \
1491      -12.4520       2.70432     -0.222601   1.116467  -0.565800   
632       -12.4057       5.60872     -2.978020  17.530006  14.007729   
4234      -10.5150       3.97760      1.072630   9.557858  12.574345   
1582      -11.7008       2.77122     -3.040940  -0.662121   2.957458   
272       -13.4821       2.90954     -0.818331   8.936436   6.064831   
1825      -13.2688       4.64589      1.143770   1.048500   3.534443   
3416       94.6584    -136.92500     82.517100   0.593174   2.757504   
936       -12.0346       3.72065     -1.634880   4.201870   3.970905   
1354      -12.2558       3.39683     -0.748180  12.009953   3.785801   
1730       -9.0751       3.75076     -1.608000   6.127354   4.417351   

      bag_orig_detrended  bag_new_detrended  
1491           -7.226678          -8.934986  
632             9.680329           6.134750  
4234            1.708180           4.701366  
1582           -8.511798          -4.915520  
272             1.086759          -1.808148  
1825           -6.801177          -4.338536  
3416           -6.763036          -4.619268  
936            -3.154340          -3.405866  
1354            4.653743          -3.590970  
1730           -1.228856          -2.959420

In [4]:
# very old
data_overview_heldout[no_nans].sort_values(by='age', ascending=False)[columns].head(10)

eid  age      bmi  digit substitution  education  \
3417  1318155   81  22.9840                15.0       20.0   
4376  1402195   81  22.8262                20.0       20.0   
3983  1366740   81  19.6259                22.0       20.0   
3085  1287410   80  25.8199                14.0       20.0   
2971  1274603   80  28.3723                14.0       20.0   
4136  1381380   80  24.4230                13.0       20.0   
4777  1440190   80  24.8922                 9.0       19.0   
3460  1321356   80  23.1978                10.0       15.0   
2514  1232260   80  21.2183                 1.0       15.0   
2611  1242018   80  24.5997                23.0       19.0   

      fluid intelligence  grip  depressive episode  all depression  \
3417                 6.0  20.0                 0.0             0.0   
4376                 7.0  35.0                 0.0             0.0   
3983                 5.0  29.0                 0.0             0.0   
3085                 8.0  26.0                 0.0             0.0   
2971                 4.0  38.0                 0.0             0.0   
4136                 5.0  20.0                 0.0             0.0   
4777                 7.0  30.0                 0.0             0.0   
3460                 9.0  18.0                 0.0             0.0   
2514                 4.0  20.0                 0.0             0.0   
2611                 9.0  40.0                 0.0             0.0   

      recurrent depressive disorder  multiple sclerosis  sex  weekly beer  \
3417                            0.0                 0.0    0          0.0   
4376                            0.0                 0.0    1          0.0   
3983                            0.0                 0.0    0          1.0   
3085                            0.0                 0.0    1          0.0   
2971                            0.0                 0.0    1          0.0   
4136                            0.0                 0.0    0          0.0   
4777                            0.0                 0.0    1          1.0   
3460                            0.0                 0.0    1          0.0   
2514                            0.0                 0.0    0          0.0   
2611                            0.0                 0.0    1          2.0   

      genetic pc 1  genetic pc 2  genetic pc 3   bag_orig    bag_new  \
3417     -11.37220      3.714020     -3.803910 -10.466156 -11.490562   
4376     -13.93000      6.717830     -2.322770  -7.379013 -10.690979   
3983     -11.07080     -0.551868     -2.593550 -15.122696 -14.865349   
3085     -13.95480      5.096500     -1.158760 -13.951309  -7.180679   
2971     -12.12880      2.780370     -1.920700  -1.807571   0.765625   
4136     -10.32720      2.258110      0.682063 -10.551620  -7.375259   
4777      -9.90560      2.722880     -2.310510  -2.085754  -4.522217   
3460     -12.05430      4.536380     -2.936910 -10.321014 -14.976639   
2514      -9.96863      2.334970      2.634670  -7.770508  -7.578476   
2611     -13.01330      2.526700     -1.493070  -8.357498  -8.505112   

      bag_orig_detrended  bag_new_detrended  
3417           -2.524873          -3.484895  
4376            0.562270          -2.685312  
3983           -7.181413          -6.859682  
3085           -6.503494           0.328780  
2971            5.640244           8.275084  
4136           -3.103805           0.134200  
4777            5.362061           2.987242  
3460           -2.873199          -7.467180  
2514           -0.322693          -0.069017  
2611           -0.909683          -0.995652

For simplicity's sake, I will choose the first-ranking subject for both the very young and very old samples:
- young: subject 1137757
- old: subject 1318155

In [5]:
# prediction much older than true age (shallow model)
data_overview_heldout[no_nans].sort_values(by='bag_orig_detrended', ascending=False)[columns].head(10)

eid  age      bmi  digit substitution  education  \
3956  1364474   71  22.0840                 6.0       20.0   
3948  1364183   63  33.8659                18.0       20.0   
4491  1412736   68  21.8944                15.0       19.0   
2770  1256413   62  20.9689                25.0       15.0   
3346  1311923   74  23.8041                18.0       15.0   
2671  1246637   53  31.1829                17.0       19.0   
1535  1141252   69  25.8150                18.0       10.0   
2807  1259547   59  28.5375                14.0       20.0   
4835  1445661   63  25.5650                27.0       20.0   
2482  1229284   74  26.2284                15.0       15.0   

      fluid intelligence  grip  depressive episode  all depression  \
3956                 7.0  30.0                 0.0             0.0   
3948                 6.0  52.0                 0.0             0.0   
4491                 7.0  22.0                 0.0             0.0   
2770                 5.0  26.0                 0.0             0.0   
3346                 5.0  26.0                 0.0             0.0   
2671                 3.0  58.0                 0.0             0.0   
1535                 3.0  11.0                 0.0             0.0   
2807                 7.0  42.0                 0.0             0.0   
4835                 7.0  38.0                 0.0             0.0   
2482                 4.0  38.0                 0.0             0.0   

      recurrent depressive disorder  multiple sclerosis  sex  weekly beer  \
3956                            0.0                 0.0    0          0.0   
3948                            0.0                 0.0    1          3.0   
4491                            0.0                 0.0    0          0.0   
2770                            0.0                 0.0    0          0.0   
3346                            0.0                 0.0    0          0.0   
2671                            0.0                 0.0    1          4.0   
1535                            0.0                 0.0    0          0.0   
2807                            0.0                 0.0    1          2.0   
4835                            0.0                 0.0    1          0.0   
2482                            0.0                 0.0    1          2.0   

      genetic pc 1  genetic pc 2  genetic pc 3   bag_orig    bag_new  \
3956      -11.5235      5.753500     -2.626010   9.223434  10.724228   
3948      -13.7643      7.592210     -3.358300  12.854942  11.279037   
4491      -12.8314      0.243051     -1.594970   9.954872   4.595406   
2770      -11.6555      4.579840      1.048380  12.504929   8.797760   
3346        6.7404      5.116490     -2.678700   6.107819   4.266815   
2671      -10.6692      4.158480     -0.772086  16.365906  12.807861   
1535      -14.4433      5.805530     -0.288483   8.446190   1.187698   
2807      -11.1963      3.047180      0.639169  13.219299  14.108459   
4835       20.4299     -9.091120     27.588400  11.107239   6.993210   
2482      -12.0258      6.017550     -1.140410   5.637833  -2.716675   

      bag_orig_detrended  bag_new_detrended  
3956           12.230042          13.767818  
3948           11.913810          10.352966  
4491           11.481077           6.150373  
2770           11.070329           7.375481  
3346           10.594829           8.799028  
2671           10.490099           6.919713  
1535           10.465863           3.238873  
2807           10.304297          11.197558  
4835           10.166107           6.067139  
2482           10.124843           1.815538

In [6]:
# prediction much older than true age (deep model)
data_overview_heldout[no_nans].sort_values(by='bag_new_detrended', ascending=False)[columns].head(10)

eid  age      bmi  digit substitution  education  \
3956  1364474   71  22.0840                 6.0       20.0   
4757  1438262   59  18.2969                16.0       20.0   
2807  1259547   59  28.5375                14.0       20.0   
4035  1371802   61  26.5118                26.0       19.0   
4849  1446985   59  30.6170                23.0       19.0   
3948  1364183   63  33.8659                18.0       20.0   
1422  1131993   61  32.0069                19.0        7.0   
4364  1401363   71  24.4919                16.0       15.0   
3374  1314229   74  23.4865                 7.0       20.0   
4879  1450407   76  25.3565                21.0       13.0   

      fluid intelligence  grip  depressive episode  all depression  \
3956                 7.0  30.0                 0.0             0.0   
4757                 4.0  20.0                 1.0             1.0   
2807                 7.0  42.0                 0.0             0.0   
4035                 6.0  36.0                 0.0             0.0   
4849                 5.0  46.0                 0.0             0.0   
3948                 6.0  52.0                 0.0             0.0   
1422                 4.0  36.0                 0.0             0.0   
4364                 9.0  40.0                 0.0             0.0   
3374                 5.0  32.0                 0.0             0.0   
4879                 8.0  16.0                 0.0             0.0   

      recurrent depressive disorder  multiple sclerosis  sex  weekly beer  \
3956                            0.0                 0.0    0          0.0   
4757                            1.0                 0.0    0          3.0   
2807                            0.0                 0.0    1          2.0   
4035                            0.0                 0.0    1          4.0   
4849                            0.0                 0.0    1          4.0   
3948                            0.0                 0.0    1          3.0   
1422                            0.0                 0.0    1          2.0   
4364                            0.0                 0.0    1          0.0   
3374                            0.0                 0.0    1          2.0   
4879                            0.0                 0.0    0          0.0   

      genetic pc 1  genetic pc 2  genetic pc 3   bag_orig    bag_new  \
3956     -11.52350       5.75350     -2.626010   9.223434  10.724228   
4757     -13.41550       1.79436      0.182334  10.040665  15.234283   
2807     -11.19630       3.04718      0.639169  13.219299  14.108459   
4035      -7.51669       2.12231     -1.098830   5.131050  12.614227   
4849     -12.01620       4.23413     -3.621510   8.651314  13.495262   
3948     -13.76430       7.59221     -3.358300  12.854942  11.279037   
1422      -7.32864      -8.90400     -0.961118  10.925354  12.192017   
4364     -14.05340       3.86783     -2.113140  -1.460297   6.956360   
3374     -11.89020       5.52711     -3.839010   0.372177   5.410660   
4879     -13.46300       3.77453     -0.483266   0.424561   4.370087   

      bag_orig_detrended  bag_new_detrended  
3956           12.230042          13.767818  
4757            7.125663          12.323381  
2807           10.304297          11.197558  
4035            3.202983          10.695741  
4849            5.736312          10.584360  
3948           11.913810          10.352966  
1422            8.997287          10.273530  
4364            1.546311           9.999950  
3374            4.859187           9.942873  
4879            5.898506           9.894715

For cases of brain ages being much higher than true ages, both models have the same number one in their ranking: subject 1364474.

In [7]:
# prediction much younger than true age (shallow model)
data_overview_heldout[no_nans].sort_values(by='bag_orig_detrended', ascending=True)[columns].head(10)

eid  age      bmi  digit substitution  education  \
1992  1184162   69  26.8905                18.0       20.0   
4160  1383475   69  24.1357                28.0       20.0   
3267  1305142   62  26.4157                27.0       19.0   
1041  1098221   52  27.1626                17.0       19.0   
3508  1325914   53  27.0527                14.0       15.0   
2902  1268821   74  26.7746                19.0       20.0   
2834  1262448   69  25.6639                22.0       20.0   
3820  1352461   75  26.0593                20.0       10.0   
1807  1167559   65  24.3844                13.0       20.0   
2386  1218873   55  29.7744                27.0       20.0   

      fluid intelligence  grip  depressive episode  all depression  \
1992                 4.0  26.0                 0.0             0.0   
4160                11.0  39.0                 0.0             0.0   
3267                 7.0  22.0                 0.0             0.0   
1041                 4.0  30.0                 0.0             0.0   
3508                 7.0  50.0                 0.0             0.0   
2902                 3.0  43.0                 0.0             0.0   
2834                 9.0  50.0                 0.0             0.0   
3820                 6.0  28.0                 0.0             0.0   
1807                 5.0  18.0                 0.0             0.0   
2386                12.0  30.0                 0.0             0.0   

      recurrent depressive disorder  multiple sclerosis  sex  weekly beer  \
1992                            0.0                 0.0    1          1.0   
4160                            0.0                 0.0    1          8.0   
3267                            0.0                 0.0    1          5.0   
1041                            0.0                 0.0    0          0.0   
3508                            0.0                 0.0    1         12.0   
2902                            0.0                 0.0    1          2.0   
2834                            0.0                 0.0    1          1.0   
3820                            0.0                 0.0    1          0.0   
1807                            0.0                 0.0    0          2.0   
2386                            0.0                 0.0    1          2.0   

      genetic pc 1  genetic pc 2  genetic pc 3   bag_orig    bag_new  \
1992     -10.53270      3.493190     -1.506880 -21.213188 -21.445992   
4160     -12.05250      3.900220     -0.600046 -14.754501  -7.280476   
3267     -14.36980     -0.042826      1.453400 -10.485474  -3.813774   
1041     -13.10390      3.904950     -4.629740  -5.303051   3.726322   
3508     -13.41250      3.418270     -3.066450  -5.414597  -5.551502   
2902     102.52200   -150.103000     81.525500 -15.238331  -8.931396   
2834     -13.33520      6.603760     -1.556730 -12.739822  -4.817749   
3820     -15.23290      5.119800     -0.732138 -15.590584 -12.477036   
1807      -8.94561      5.527930      0.215461 -10.603779 -10.705269   
2386     -11.04390      4.403990     -0.450628  -5.211079   0.442238   

      bag_orig_detrended  bag_new_detrended  
1992          -19.193515         -19.394817  
4160          -12.734829          -5.229301  
3267          -11.920073          -5.236053  
1041          -11.672326          -2.658034  
3508          -11.290404         -11.439650  
2902          -10.751321          -4.399183  
2834          -10.720150          -2.766574  
3820          -10.610106          -7.448615  
1807          -10.557976         -10.638925  
2386          -10.099951          -4.453495

In [8]:
# prediction much younger than true age (deep model)
data_overview_heldout[no_nans].sort_values(by='bag_new_detrended', ascending=True)[columns].head(10)

eid  age      bmi  digit substitution  education  \
1992  1184162   69  26.8905                18.0       20.0   
4864  1448799   66  23.5482                15.0       20.0   
2787  1258107   61  27.2151                17.0       15.0   
3508  1325914   53  27.0527                14.0       15.0   
1807  1167559   65  24.3844                13.0       20.0   
2786  1257964   59  28.1634                19.0       20.0   
3899  1359852   50  28.4665                26.0       19.0   
4056  1374443   55  22.6912                27.0       20.0   
2448  1225260   68  28.8493                19.0       20.0   
148   1014619   54  30.5623                32.0       20.0   

      fluid intelligence  grip  depressive episode  all depression  \
1992                 4.0  26.0                 0.0             0.0   
4864                 8.0  12.0                 0.0             0.0   
2787                 4.0  48.0                 0.0             0.0   
3508                 7.0  50.0                 0.0             0.0   
1807                 5.0  18.0                 0.0             0.0   
2786                 4.0   8.0                 0.0             0.0   
3899                 7.0  32.0                 0.0             0.0   
4056                 7.0  28.0                 0.0             0.0   
2448                 5.0  40.0                 0.0             0.0   
148                  7.0  27.0                 0.0             0.0   

      recurrent depressive disorder  multiple sclerosis  sex  weekly beer  \
1992                            0.0                 0.0    1          1.0   
4864                            0.0                 0.0    0          2.0   
2787                            0.0                 0.0    1          2.0   
3508                            0.0                 0.0    1         12.0   
1807                            0.0                 0.0    0          2.0   
2786                            0.0                 0.0    0          1.0   
3899                            0.0                 0.0    0          0.0   
4056                            0.0                 0.0    0          0.0   
2448                            0.0                 0.0    1          0.0   
148                             0.0                 0.0    0          0.0   

      genetic pc 1  genetic pc 2  genetic pc 3   bag_orig    bag_new  \
1992     -10.53270       3.49319     -1.506880 -21.213188 -21.445992   
4864     -12.62460       4.20173     -2.959480  -8.648052 -13.540424   
2787     -10.08450       3.67660     -1.570830  -7.522968 -10.298618   
3508     -13.41250       3.41827     -3.066450  -5.414597  -5.551502   
1807      -8.94561       5.52793      0.215461 -10.603779 -10.705269   
2786     -11.43140       3.77235     -0.585954  -2.409645  -7.383686   
3899     -11.24260       1.55585     -3.185640  -1.955082  -2.548946   
4056     -11.11600       6.15412     -4.098720  -1.009426  -5.005344   
2448     -10.82400       4.20293     -0.249413 -10.676670 -11.013927   
148      -12.92810       3.04781     -0.786850   3.897453  -3.618961   

      bag_orig_detrended  bag_new_detrended  
1992          -19.193515         -19.394817  
4864           -8.108782         -12.977873  
2787           -9.451035         -12.217105  
3508          -11.290404         -11.439650  
1807          -10.557976         -10.638925  
2786           -5.324647         -10.294588  
3899           -9.311291          -9.925717  
4056           -5.898298          -9.901077  
2448           -9.150465          -9.458960  
148            -1.484886          -9.010902

For cases of brain ages being much lower than true ages, both models have the same number one in their ranking: subject 1184162.

In [9]:
# get difference between models
data_overview_heldout['bag diff'] = (data_overview_heldout['bag_orig'] - data_overview_heldout['bag_new']).abs()
data_overview_heldout['bag diff detrended'] = (data_overview_heldout['bag_orig_detrended'] - data_overview_heldout['bag_new_detrended']).abs()
data_overview_heldout

eid  age         split      bmi  digit substitution  education  \
0     1000030   65  heldout_test  21.2569                 NaN       20.0   
1     1000059   68  heldout_test      NaN                22.0       20.0   
2     1000077   58  heldout_test  33.4834                18.0       19.0   
3     1000100   57  heldout_test  23.4552                 NaN       20.0   
4     1000202   67  heldout_test  26.6846                19.0       20.0   
...       ...  ...           ...      ...                 ...        ...   
4995  2035762   73  heldout_test  25.3688                12.0        7.0   
4996  2264650   61  heldout_test  30.5389                 NaN        7.0   
4997  2329354   68  heldout_test  22.5614                18.0       20.0   
4998  2457356   60  heldout_test  29.6270                18.0       20.0   
4999  2584074   49  heldout_test  21.8522                 NaN       20.0   

      fluid intelligence  grip  depressive episode  all depression  ...  \
0                    8.0  18.0                 1.0             1.0  ...   
1                    7.0   NaN                 0.0             0.0  ...   
2                    3.0  48.0                 NaN             NaN  ...   
3                    NaN  28.0                 0.0             0.0  ...   
4                    6.0  45.0                 NaN             NaN  ...   
...                  ...   ...                 ...             ...  ...   
4995                 5.0  24.0                 0.0             0.0  ...   
4996                 NaN  41.0                 0.0             0.0  ...   
4997                 8.0  30.0                 0.0             0.0  ...   
4998                 8.0  22.0                 1.0             1.0  ...   
4999                11.0  64.0                 1.0             1.0  ...   

      batch_nb_orig  predicted_age_orig  batch_nb_new  predicted_age_new  \
0               1.0           64.256958           1.0          71.693298   
1              18.0           58.876446          18.0          58.413288   
2              34.0           63.836048          34.0          66.193604   
3               2.0           55.072990           2.0          57.406872   
4              16.0           64.058601          16.0          67.427048   
...             ...                 ...           ...                ...   
4995            7.0           77.852402           7.0          74.190857   
4996           29.0           61.950256          29.0          63.557865   
4997           36.0           66.802460          36.0          65.315231   
4998            9.0           61.610027           9.0          63.635746   
4999           12.0           60.643711          12.0          57.262020   

       bag_orig   bag_new  bag_orig_detrended  bag_new_detrended  bag diff  \
0     -0.743042  6.693298           -0.697239           6.759642  7.436340   
1     -9.123554 -9.586712           -7.597349          -8.031745  0.463158   
2      5.836048  8.193604            2.427579           4.786494  2.357555   
3     -1.927010  0.406872           -5.828947          -3.496446  2.333881   
4     -2.941399  0.427048           -1.908661           1.485807  3.368446   
...         ...       ...                 ...                ...       ...   
4995   4.852402  1.190857            8.845944           5.226862  3.661545   
4996   0.950256  2.557865           -0.977811           0.639379  1.607609   
4997  -1.197540 -2.684769            0.328665          -1.129802  1.487228   
4998   1.610027  3.635746           -0.811507           1.221052  2.025719   
4999  11.643711  8.262020            3.794034           0.389041  3.381691   

      bag diff detrended  
0               7.456882  
1               0.434396  
2               2.358915  
3               2.332501  
4               3.394468  
...                  ...  
4995            3.619082  
4996            1.617189  
4997            1.458467  
4998            2.032559  
4999            3.404993  

[5000 rows x 27 colum

In [10]:
# max difference
data_overview_heldout[no_nans].sort_values(by='bag diff', ascending=False)[columns+['bag diff','bag diff detrended']].head(10)

eid  age      bmi  digit substitution  education  \
4968  1457519   59  24.0490                26.0       20.0   
1041  1098221   52  27.1626                17.0       19.0   
4364  1401363   71  24.4919                16.0       15.0   
2482  1229284   74  26.2284                15.0       15.0   
3788  1350106   74  26.3565                16.0       19.0   
1354  1125280   50  32.8291                25.0       15.0   
3958  1364685   61  29.4650                24.0       13.0   
4665  1429687   62  29.5098                24.0       20.0   
4004  1369645   77  29.5156                22.0       19.0   
2834  1262448   69  25.6639                22.0       20.0   

      fluid intelligence  grip  depressive episode  all depression  \
4968                 9.0  48.0                 0.0             0.0   
1041                 4.0  30.0                 0.0             0.0   
4364                 9.0  40.0                 0.0             0.0   
2482                 4.0  38.0                 0.0             0.0   
3788                 4.0  34.0                 0.0             0.0   
1354                 5.0  53.0                 0.0             0.0   
3958                 8.0  42.0                 0.0             0.0   
4665                 7.0  31.0                 0.0             0.0   
4004                 4.0  33.0                 0.0             0.0   
2834                 9.0  50.0                 0.0             0.0   

      recurrent depressive disorder  ...  weekly beer  genetic pc 1  \
4968                            0.0  ...          6.0      -12.7438   
1041                            0.0  ...          0.0      -13.1039   
4364                            0.0  ...          0.0      -14.0534   
2482                            0.0  ...          2.0      -12.0258   
3788                            0.0  ...          6.0      -15.5866   
1354                            0.0  ...          3.0      -12.2558   
3958                            0.0  ...          3.0      -10.9326   
4665                            0.0  ...          6.0      -11.6891   
4004                            0.0  ...          0.0      -12.6389   
2834                            0.0  ...          1.0      -13.3352   

      genetic pc 2  genetic pc 3   bag_orig   bag_new  bag_orig_detrended  \
4968       4.27521     -2.563990  12.981110  1.826702           10.066107   
1041       3.90495     -4.629740  -5.303051  3.726322          -11.672326   
4364       3.86783     -2.113140  -1.460297  6.956360            1.546311   
2482       6.01755     -1.140410   5.637833 -2.716675           10.124843   
3788       3.28219     -3.866920  -8.842018 -0.556030           -4.355008   
1354       3.39683     -0.748180  12.009953  3.785801            4.653743   
3958       5.01633     -3.363310  -2.206276  5.748245           -4.134343   
4665       1.53879     -0.974370  -3.541058  4.400703           -4.975657   
4004       4.12183     -0.904572   3.673676 -4.265297            9.641088   
2834       6.60376     -1.556730 -12.739822 -4.817749          -10.720150   

      bag_new_detrended   bag diff  bag diff detrended  
4968          -1.084200  11.154408           11.150307  
1041          -2.658034   9.029373            9.014292  
4364           9.999950   8.416656            8.453639  
2482           1.815538   8.354507            8.309304  
3788           3.976183   8.285988            8.331191  
1354          -3.590970   8.224152            8.244713  
3958           3.829759   7.954521            7.964102  
4665           2.978425   7.941761            7.954082  
4004           1.755539   7.938972            7.885549  
2834          -2.766574   7.922073            7.953575  

[10 rows x 22 columns]

In [11]:
# max difference (detrended)
data_overview_heldout[no_nans].sort_values(by='bag diff detrended', ascending=False)[columns+['bag diff','bag diff detrended']].head(10)

eid  age      bmi  digit substitution  education  \
4968  1457519   59  24.0490                26.0       20.0   
1041  1098221   52  27.1626                17.0       19.0   
4364  1401363   71  24.4919                16.0       15.0   
3788  1350106   74  26.3565                16.0       19.0   
2482  1229284   74  26.2284                15.0       15.0   
1354  1125280   50  32.8291                25.0       15.0   
3958  1364685   61  29.4650                24.0       13.0   
4665  1429687   62  29.5098                24.0       20.0   
2834  1262448   69  25.6639                22.0       20.0   
4004  1369645   77  29.5156                22.0       19.0   

      fluid intelligence  grip  depressive episode  all depression  \
4968                 9.0  48.0                 0.0             0.0   
1041                 4.0  30.0                 0.0             0.0   
4364                 9.0  40.0                 0.0             0.0   
3788                 4.0  34.0                 0.0             0.0   
2482                 4.0  38.0                 0.0             0.0   
1354                 5.0  53.0                 0.0             0.0   
3958                 8.0  42.0                 0.0             0.0   
4665                 7.0  31.0                 0.0             0.0   
2834                 9.0  50.0                 0.0             0.0   
4004                 4.0  33.0                 0.0             0.0   

      recurrent depressive disorder  ...  weekly beer  genetic pc 1  \
4968                            0.0  ...          6.0      -12.7438   
1041                            0.0  ...          0.0      -13.1039   
4364                            0.0  ...          0.0      -14.0534   
3788                            0.0  ...          6.0      -15.5866   
2482                            0.0  ...          2.0      -12.0258   
1354                            0.0  ...          3.0      -12.2558   
3958                            0.0  ...          3.0      -10.9326   
4665                            0.0  ...          6.0      -11.6891   
2834                            0.0  ...          1.0      -13.3352   
4004                            0.0  ...          0.0      -12.6389   

      genetic pc 2  genetic pc 3   bag_orig   bag_new  bag_orig_detrended  \
4968       4.27521     -2.563990  12.981110  1.826702           10.066107   
1041       3.90495     -4.629740  -5.303051  3.726322          -11.672326   
4364       3.86783     -2.113140  -1.460297  6.956360            1.546311   
3788       3.28219     -3.866920  -8.842018 -0.556030           -4.355008   
2482       6.01755     -1.140410   5.637833 -2.716675           10.124843   
1354       3.39683     -0.748180  12.009953  3.785801            4.653743   
3958       5.01633     -3.363310  -2.206276  5.748245           -4.134343   
4665       1.53879     -0.974370  -3.541058  4.400703           -4.975657   
2834       6.60376     -1.556730 -12.739822 -4.817749          -10.720150   
4004       4.12183     -0.904572   3.673676 -4.265297            9.641088   

      bag_new_detrended   bag diff  bag diff detrended  
4968          -1.084200  11.154408           11.150307  
1041          -2.658034   9.029373            9.014292  
4364           9.999950   8.416656            8.453639  
3788           3.976183   8.285988            8.331191  
2482           1.815538   8.354507            8.309304  
1354          -3.590970   8.224152            8.244713  
3958           3.829759   7.954521            7.964102  
4665           2.978425   7.941761            7.954082  
2834          -2.766574   7.922073            7.953575  
4004           1.755539   7.938972            7.885549  

[10 rows x 22 columns]

Looking at the maximum BAG differences between model predictions, both the "original" and the detrended BAGs result in almost identical top 10 rankings (only 9 and 10 are swapped). The highest ranking sub in both cases is: subject 1457519

In [12]:
# min difference 
data_overview_heldout[no_nans].sort_values(by='bag diff', ascending=True)[columns+['bag diff','bag diff detrended']].head(10)

eid  age      bmi  digit substitution  education  \
332   1031078   58  30.9633                24.0       19.0   
3414  1317822   68  27.4691                15.0       10.0   
4235  1389218   60  23.7582                17.0       15.0   
3962  1364956   72  24.8771                14.0       20.0   
4900  1451946   73  30.7107                10.0       20.0   
4641  1426814   64  24.2209                14.0       20.0   
4096  1377640   52  30.0085                21.0       20.0   
3351  1312176   79  27.7254                 6.0       15.0   
4135  1381322   79  23.6577                16.0       19.0   
4904  1452267   69  29.1416                17.0       20.0   

      fluid intelligence  grip  depressive episode  all depression  \
332                  8.0  40.0                 0.0             0.0   
3414                 6.0  40.0                 0.0             0.0   
4235                 7.0  24.0                 0.0             0.0   
3962                 6.0  36.0                 0.0             0.0   
4900                 6.0  46.0                 0.0             0.0   
4641                 7.0  20.0                 0.0             0.0   
4096                 8.0  56.0                 0.0             0.0   
3351                 5.0  27.0                 0.0             0.0   
4135                 7.0  40.0                 0.0             0.0   
4904                 7.0  47.0                 0.0             0.0   

      recurrent depressive disorder  ...  weekly beer  genetic pc 1  \
332                             0.0  ...          2.0      -9.16589   
3414                            0.0  ...          1.0     -13.14560   
4235                            0.0  ...          0.0     -12.25090   
3962                            0.0  ...         15.0     -12.75630   
4900                            0.0  ...          0.0     -12.12510   
4641                            0.0  ...          2.0     -12.62990   
4096                            0.0  ...          4.0     -10.52980   
3351                            0.0  ...         10.0     -13.83750   
4135                            0.0  ...          0.0     -10.74030   
4904                            0.0  ...          0.0     -13.84000   

      genetic pc 2  genetic pc 3   bag_orig    bag_new  bag_orig_detrended  \
332        3.51011     -1.576750  12.719391  12.724068            9.310921   
3414       6.21674     -1.552870   1.406654   1.411583            2.932860   
4235       4.30612     -1.309870   3.437328   3.442513            1.015794   
3962       3.48912     -2.582970  -2.894646  -2.889091            0.605430   
4900       4.03849     -3.847240  -0.905197  -0.910950            3.088346   
4641       3.64182     -2.328940  -1.163151  -1.170139           -1.610815   
4096       2.66272     -0.873506   6.302193   6.312489           -0.067082   
3351       4.01783     -3.643330  -4.747574  -4.758247            2.206774   
4135       3.27624     -1.711470  -4.833069  -4.846535            2.121279   
4904       2.13743     -3.619880  -6.401356  -6.416119           -4.381683   

      bag_new_detrended  bag diff  bag diff detrended  
332            9.316958  0.004677            0.006037  
3414           2.966550  0.004929            0.033690  
4235           1.027818  0.005184            0.012025  
3962           0.650706  0.005554            0.045277  
4900           3.125056  0.005753            0.036710  
4641          -1.600003  0.006989            0.010813  
4096          -0.071867  0.010296            0.004785  
3351           2.255004  0.010674            0.048230  
4135           2.166717  0.013466            0.045438  
4904          -4.364944  0.014763            0.016739  

[10 rows x 22 columns]

In [13]:
# min difference (detrended)
data_overview_heldout[no_nans].sort_values(by='bag diff detrended', ascending=True)[columns+['bag diff','bag diff detrended']].head(10)

eid  age      bmi  digit substitution  education  \
796   1075092   67  22.1384                17.0       10.0   
4096  1377640   52  30.0085                21.0       20.0   
332   1031078   58  30.9633                24.0       19.0   
4648  1427566   78  23.5177                16.0       10.0   
1053  1099607   67  24.7556                19.0       20.0   
4641  1426814   64  24.2209                14.0       20.0   
4235  1389218   60  23.7582                17.0       15.0   
3114  1290223   71  30.7174                22.0       20.0   
4904  1452267   69  29.1416                17.0       20.0   
4487  1412450   61  24.2969                25.0       19.0   

      fluid intelligence  grip  depressive episode  all depression  \
796                  4.0  20.0                 0.0             0.0   
4096                 8.0  56.0                 0.0             0.0   
332                  8.0  40.0                 0.0             0.0   
4648                 7.0  30.0                 0.0             0.0   
1053                 9.0  27.0                 0.0             0.0   
4641                 7.0  20.0                 0.0             0.0   
4235                 7.0  24.0                 0.0             0.0   
3114                 8.0  42.0                 0.0             0.0   
4904                 7.0  47.0                 0.0             0.0   
4487                 6.0  20.0                 0.0             0.0   

      recurrent depressive disorder  ...  weekly beer  genetic pc 1  \
796                             0.0  ...          0.0     -12.43780   
4096                            0.0  ...          4.0     -10.52980   
332                             0.0  ...          2.0      -9.16589   
4648                            0.0  ...          3.0     -10.85980   
1053                            0.0  ...          0.0     -14.28630   
4641                            0.0  ...          2.0     -12.62990   
4235                            0.0  ...          0.0     -12.25090   
3114                            0.0  ...          2.0     -12.77980   
4904                            0.0  ...          0.0     -13.84000   
4487                            0.0  ...          0.0     -15.25810   

      genetic pc 2  genetic pc 3   bag_orig    bag_new  bag_orig_detrended  \
796        3.29044     -0.210425  -5.625736  -5.654293           -4.592998   
4096       2.66272     -0.873506   6.302193   6.312489           -0.067082   
332        3.51011     -1.576750  12.719391  12.724068            9.310921   
4648       3.29852     -3.487720  -8.367172  -8.415024           -1.906292   
1053       4.82959     -3.587270   3.678940   3.642349            4.711678   
4641       3.64182     -2.328940  -1.163151  -1.170139           -1.610815   
4235       4.30612     -1.309870   3.437328   3.442513            1.015794   
3114       4.53645     -3.550240  -6.295181  -6.345970           -3.288574   
4904       2.13743     -3.619880  -6.401356  -6.416119           -4.381683   
4487       3.02657     -1.604460   3.772217   3.744255            1.844150   

      bag_new_detrended  bag diff  bag diff detrended  
796           -4.595534  0.028557            0.002535  
4096          -0.071867  0.010296            0.004785  
332            9.316958  0.004677            0.006037  
4648          -1.897980  0.047852            0.008312  
1053           4.701109  0.036591            0.010569  
4641          -1.600003  0.006989            0.010813  
4235           1.027818  0.005184            0.012025  
3114          -3.302380  0.050789            0.013806  
4904          -4.364944  0.014763            0.016739  
4487           1.825768  0.027962            0.018381  

[10 rows x 22 columns]

The lists for minimal BAG differences across model predictions is more varied between the "original" and detrended BAGs than the maximal differences viewed before. In both lists, one subject comes in the top 3: subject 1031078